File: neurons/execution_layer/proof_uploader.py
import base64
import json
import logging
import os
from typing import Optional

import httpx

import cli_parser

logger = logging.getLogger(__name__)

SN2_API_URL = os.getenv("SN2_API_URL", "https://sn2-api.inferencelabs.com")


def _get_api_url() -> str:
    return getattr(cli_parser.config, "sn2_api_url", None) or SN2_API_URL


def _get_wallet():
    return cli_parser.config.wallet


def _sign_body(body: str) -> str:
    wallet = _get_wallet()
    signature = wallet.hotkey.sign(body.encode())
    return base64.b64encode(signature).decode()


def _authenticated_post(url: str, payload: dict, timeout: float = 60.0) -> dict:
    body = json.dumps(payload)
    signature = _sign_body(body)
    with httpx.Client(timeout=timeout) as client:
        response = client.post(
            url,
            content=body,
            headers={
                "Content-Type": "application/json",
                "X-Request-Signature": signature,
            },
        )
        response.raise_for_status()
        return response.json()


def request_upload_urls(
    run_uid: str,
    circuit_id: str,
    artifacts: list[dict],
) -> list[dict]:
    api_url = _get_api_url()
    hotkey = _get_wallet().hotkey.ss58_address
    payload = {
        "validator_key": hotkey,
        "run_uid": run_uid,
        "circuit_id": circuit_id,
        "artifacts": artifacts,
    }
    result = _authenticated_post(f"{api_url}/proofs/upload-urls", payload)
    return result.get("urls", [])


def upload_artifact_bytes(client: httpx.Client, upload_url: str, data: bytes) -> bool:
    response = client.put(
        upload_url,
        content=data,
        headers={"Content-Type": "application/octet-stream"},
    )
    return response.status_code in (200, 201)


def confirm_uploads(
    run_uid: str,
    circuit_id: str,
    circuit_name: str,
    artifacts: list[dict],
) -> dict:
    api_url = _get_api_url()
    hotkey = _get_wallet().hotkey.ss58_address
    payload = {
        "validator_key": hotkey,
        "run_uid": run_uid,
        "circuit_id": circuit_id,
        "circuit_name": circuit_name,
        "artifacts": artifacts,
    }
    return _authenticated_post(f"{api_url}/proofs/confirm", payload)


def upload_final_output(
    run_uid: str,
    circuit_id: str,
    output: dict,
) -> Optional[str]:
    api_url = _get_api_url()
    hotkey = _get_wallet().hotkey.ss58_address
    payload = {
        "validator_key": hotkey,
        "run_uid": run_uid,
        "circuit_id": circuit_id,
        "output": output,
    }
    result = _authenticated_post(f"{api_url}/proofs/{run_uid}/output", payload)
    return result.get("gcs_key")


def upload_run_proofs(
    run_uid: str,
    circuit_id: str,
    circuit_name: str,
    proof_artifacts: list[dict],
    final_output: Optional[dict] = None,
) -> bool:
    """
    Upload all proofs for a completed run via signed URLs.

    proof_artifacts: list of dicts with keys:
        - slice_num: str
        - proof_system: str
        - proof_data: bytes | str (hex for JSTPROVE, dict for EZKL)
        - parent_slice: Optional[str]
        - tile_idx: Optional[int]
    """
    try:
        hotkey = _get_wallet().hotkey.ss58_address
        if hotkey == "unknown":
            logger.warning(f"Skipping proof upload for run {run_uid}: invalid hotkey")
            return False
    except Exception:
        logger.warning(f"Skipping proof upload for run {run_uid}: wallet not available")
        return False

    if not proof_artifacts:
        logger.info(f"No proof artifacts to upload for run {run_uid}")
        return True

    try:
        artifact_specs = [
            {"slice_num": a["slice_num"], "artifact_type": "proof"}
            for a in proof_artifacts
        ]
    except (KeyError, TypeError) as e:
        logger.error(f"Malformed proof artifact for run {run_uid}: {e}")
        return False

    try:
        url_responses = request_upload_urls(run_uid, circuit_id, artifact_specs)
    except Exception as e:
        logger.error(f"Failed to request upload URLs for run {run_uid}: {e}")
        return False

    try:
        url_map = {(u["slice_num"], u["artifact_type"]): u for u in url_responses}
    except (KeyError, TypeError) as e:
        logger.error(f"Malformed upload URL response for run {run_uid}: {e}")
        return False

    confirm_artifacts = []
    ok = True
    with httpx.Client(timeout=120.0) as client:
        for artifact in proof_artifacts:
            try:
                key = (artifact["slice_num"], "proof")
                url_info = url_map.get(key)
                if not url_info:
                    logger.warning(f"No upload URL for {key}")
                    ok = False
                    continue

                proof_data = artifact["proof_data"]
                if isinstance(proof_data, str):
                    data_bytes = bytes.fromhex(proof_data)
                elif isinstance(proof_data, dict):
                    data_bytes = json.dumps(proof_data).encode()
                elif isinstance(proof_data, bytes):
                    data_bytes = proof_data
                else:
                    logger.warning(
                        f"Unknown proof data type for {key}: {type(proof_data)}"
                    )
                    ok = False
                    continue

                uploaded = upload_artifact_bytes(
                    client, url_info["upload_url"], data_bytes
                )
                if not uploaded:
                    logger.warning(f"Failed to upload artifact {key}")
                    ok = False
                    continue

                confirm_artifacts.append(
                    {
                        "slice_num": artifact["slice_num"],
                        "parent_slice": artifact.get("parent_slice"),
                        "tile_idx": artifact.get("tile_idx"),
                        "proof_system": artifact["proof_system"],
                        "gcs_key": url_info["gcs_key"],
                        "size_bytes": len(data_bytes),
                        "artifact_type": "proof",
                    }
                )
            except Exception as e:
                logger.error(
                    f"Error processing artifact {artifact.get('slice_num', '?')}: {e}"
                )
                ok = False
                continue

    if confirm_artifacts:
        try:
            confirm_uploads(run_uid, circuit_id, circuit_name, confirm_artifacts)
            logger.info(
                f"Confirmed {len(confirm_artifacts)} artifacts for run {run_uid}"
            )
        except Exception as e:
            logger.error(f"Failed to confirm uploads for run {run_uid}: {e}")
            ok = False

    if final_output:
        try:
            upload_final_output(run_uid, circuit_id, final_output)
            logger.info(f"Uploaded final output for run {run_uid}")
        except Exception as e:
            logger.error(f"Failed to upload final output for run {run_uid}: {e}")
            ok = False

    return ok

File: crates/sn2-validator/src/proof_uploader.rs
use std::sync::Arc;

use anyhow::{Context, Result};
use base64::Engine;
use sn2_chain::Wallet;
use tracing::{info, warn};

use crate::incremental_runner::SliceArtifact;

const DEFAULT_API_URL: &str = "https://sn2-api.inferencelabs.com";

pub struct ProofUploader {
    http: reqwest::Client,
    wallet: Arc<Wallet>,
    api_base_url: String,
}

impl ProofUploader {
    pub fn new(wallet: Arc<Wallet>, api_base_url: Option<String>) -> Self {
        Self {
            http: reqwest::Client::builder()
                .timeout(std::time::Duration::from_secs(60))
                .build()
                .unwrap_or_default(),
            wallet,
            api_base_url: api_base_url.unwrap_or_else(|| DEFAULT_API_URL.to_string()),
        }
    }

    fn sign_body(&self, body: &[u8]) -> Result<String> {
        let sig_bytes = self.wallet.sign_hotkey(body)?;
        Ok(base64::engine::general_purpose::STANDARD.encode(&sig_bytes))
    }

    pub async fn upload_run_artifacts(
        &self,
        run_uid: &str,
        circuit_id: &str,
        circuit_name: &str,
        artifacts: Vec<SliceArtifact>,
        final_output: Option<serde_json::Value>,
    ) -> Result<()> {
        if artifacts.is_empty() {
            return Ok(());
        }

        let artifact_metas: Vec<serde_json::Value> = artifacts
            .iter()
            .map(|a| {
                serde_json::json!({
                    "slice_num": a.slice_num,
                    "artifact_type": "proof",
                })
            })
            .collect();

        let urls = self
            .request_upload_urls(run_uid, circuit_id, &artifact_metas)
            .await?;

        let mut confirmed = Vec::new();

        for (artifact, url_entry) in artifacts.iter().zip(urls.iter()) {
            let upload_url = match url_entry.get("upload_url").and_then(|v| v.as_str()) {
                Some(u) => u,
                None => continue,
            };
            let gcs_key = url_entry
                .get("gcs_key")
                .and_then(|v| v.as_str())
                .unwrap_or_default();

            let proof_bytes = match &artifact.proof_hex {
                Some(hex_str) => match hex::decode(hex_str) {
                    Ok(bytes) => bytes,
                    Err(e) => {
                        warn!(slice = %artifact.slice_num, error = %e, "hex decode failed, skipping artifact");
                        continue;
                    }
                },
                None => continue,
            };

            if let Err(e) = self.upload_artifact(upload_url, &proof_bytes).await {
                warn!(
                    slice = %artifact.slice_num,
                    error = %e,
                    "artifact upload failed, skipping"
                );
                continue;
            }

            confirmed.push(serde_json::json!({
                "slice_num": artifact.slice_num,
                "proof_system": artifact.proof_system.as_ref().map(|ps| ps.to_string()),
                "gcs_key": gcs_key,
                "size_bytes": proof_bytes.len(),
                "tile_idx": artifact.tile_idx,
                "artifact_type": "proof",
            }));
        }

        if !confirmed.is_empty() {
            self.confirm_uploads(run_uid, circuit_id, circuit_name, confirmed)
                .await?;
        }

        if let Some(output) = final_output {
            self.upload_final_output(run_uid, circuit_id, &output)
                .await?;
        }

        info!(run_uid = %run_uid, "proof artifacts uploaded");
        Ok(())
    }

    async fn request_upload_urls(
        &self,
        run_uid: &str,
        circuit_id: &str,
        artifacts: &[serde_json::Value],
    ) -> Result<Vec<serde_json::Value>> {
        let body = serde_json::json!({
            "validator_key": self.wallet.hotkey_ss58(),
            "run_uid": run_uid,
            "circuit_id": circuit_id,
            "artifacts": artifacts,
        });

        let body_bytes = serde_json::to_vec(&body)?;
        let sig = self.sign_body(&body_bytes)?;

        let resp = self
            .http
            .post(format!("{}/proofs/upload-urls", self.api_base_url))
            .header("Content-Type", "application/json")
            .header("X-Request-Signature", sig)
            .body(body_bytes)
            .send()
            .await
            .context("requesting upload URLs")?;

        let status = resp.status();
        let text = resp.text().await.unwrap_or_default();
        if !status.is_success() {
            anyhow::bail!("upload-urls returned {status}: {text}");
        }

        let parsed: serde_json::Value = serde_json::from_str(&text)?;
        let urls = parsed
            .get("urls")
            .or_else(|| parsed.get("artifacts"))
            .and_then(|v| v.as_array())
            .cloned()
            .unwrap_or_default();

        Ok(urls)
    }

    async fn upload_artifact(&self, upload_url: &str, data: &[u8]) -> Result<()> {
        let resp = self
            .http
            .put(upload_url)
            .header("Content-Type", "application/octet-stream")
            .body(data.to_vec())
            .send()
            .await
            .context("uploading artifact")?;

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            anyhow::bail!("artifact PUT returned {status}: {text}");
        }
        Ok(())
    }

    async fn confirm_uploads(
        &self,
        run_uid: &str,
        circuit_id: &str,
        circuit_name: &str,
        confirmed: Vec<serde_json::Value>,
    ) -> Result<()> {
        let body = serde_json::json!({
            "validator_key": self.wallet.hotkey_ss58(),
            "run_uid": run_uid,
            "circuit_id": circuit_id,
            "circuit_name": circuit_name,
            "artifacts": confirmed,
        });

        let body_bytes = serde_json::to_vec(&body)?;
        let sig = self.sign_body(&body_bytes)?;

        let resp = self
            .http
            .post(format!("{}/proofs/confirm", self.api_base_url))
            .header("Content-Type", "application/json")
            .header("X-Request-Signature", sig)
            .body(body_bytes)
            .send()
            .await
            .context("confirming uploads")?;

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            anyhow::bail!("confirm_uploads returned {status}: {text}");
        }
        Ok(())
    }

    async fn upload_final_output(
        &self,
        run_uid: &str,
        circuit_id: &str,
        output: &serde_json::Value,
    ) -> Result<()> {
        let body = serde_json::json!({
            "validator_key": self.wallet.hotkey_ss58(),
            "run_uid": run_uid,
            "circuit_id": circuit_id,
            "output": output,
        });

        let body_bytes = serde_json::to_vec(&body)?;
        let sig = self.sign_body(&body_bytes)?;

        let resp = self
            .http
            .post(format!("{}/proofs/{}/output", self.api_base_url, run_uid))
            .header("Content-Type", "application/json")
            .header("X-Request-Signature", sig)
            .body(body_bytes)
            .send()
            .await
            .context("uploading final output")?;

        if !resp.status().is_success() {
            let status = resp.status();
            let text = resp.text().await.unwrap_or_default();
            anyhow::bail!("upload_final_output returned {status}: {text}");
        }
        Ok(())
    }
}
